plan/optimizer.go

```diff
@@ -54,6 +55,34 @@ type logicalOptRule interface {
 	optimize(LogicalPlan, context.Context, *idAllocator) (LogicalPlan, error)
 }
 
+// OptimizePreparedStmt allows us to reuse the compiled plan of a prepared statement in the same session.
+func OptimizePreparedStmt(ctx context.Context, is infoschema.InfoSchema, pstmtID uint32, schemaVer int64, node ast.Node, useCache bool) (Plan, error) {
+	var key string
+
+	ctx.GetSessionVars().StmtCtx.UseCache = useCache
+	if useCache {
+		key = key4PreparedStmt(ctx, pstmtID, schemaVer)
+		if cacheEntry, exists := ctx.Plan().Get(key); exists {
+			return cacheEntry.(Plan), nil
+		}
+	}
+	p, err := Optimize(ctx, node, is)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	if useCache {
+		ctx.Plan().Set(key, p.(interface{}))
+	}
+
+	return p, err
+}
+
+// key4PreparedStmt converts prepared statement ID with its context to a formatted string
+func key4PreparedStmt(ctx context.Context, pstmtID uint32, schemaVer int64) string {
+	return ctx.GetSessionVars().CurrentDB + ":" + strconv.FormatUint(ctx.GetSessionVars().ConnectionID, 10) + ":" +
```

Comment:
can we remove `ctx.GetSessionVars().CurrentDB` from key ?

Reply:
No, we can’t because the physical plans are dependent on the current database when building them.